### PR TITLE
add "burst" validation check to code-generator 

### PR DESCRIPTION
**Pull request by yuyulei**
**What type of PR is this?**

> /kind bug

**What this PR does / why we need it**:
remove the generation of ratelimiter in apiextensions-apiserver 's `NewForConfig`,  for two reasons:
1. no protecting for valid value of `Burst`. When using  apiextensions-apiserver 's `NewForConfig`, assume that setting config.QPS=10 but no setting config.Burst, so default value is 0, will create ratelimiter with (qps:10. burst:0). It's not expected.
2. apiextensions-apiserver 's `NewForConfig` ==>(call) 
    "k8s.io/client-go/rest/config.go" 's `RESTClientFor` ==>(cal)
    "k8s.io/client-go/rest/client.go" 's `NewRESTClient`, will new ratelimter, like:
```
if maxQPS > 0 && rateLimiter == nil {
	throttle = flowcontrol.NewTokenBucketRateLimiter(maxQPS, maxBurst)
}
```
So  no need to new ratelimiter in `apiextensions-apiserver` 's `NewForConfig`.
And why check maxQPS > 0 but no maxBurst > 0?

**Which issue(s) this PR fixes**:
no

**Special notes for your reviewer**:
no

**Does this PR introduce a user-facing change?**:
no

```release-note
NONE
```

### CODE REVIEW

**Comment by yue9944882** [comment]
could you expand what you mean by `no protecting`? the client-side qps/burst ratelimit for loopback clients is set to 50/100 on kube-apiserver's launching flow:

[https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go#L44-L45](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/config_selfclient.go#L44-L45)

and the proposed changes is in the generated code hence invalid.

**Comment by yuyulei** [comment]
@yue9944882 
 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/clientset.go#L57, 
here no check if c.Burst > 0. 

if config.QPS is set, but Burst is not set(default is 0)  will result in 

```
clientset = NewTokenBucketRateLimiter (qps, 0)
```
then the clientset could not send any request.

**Comment by yue9944882** [comment]
mmm.. the code is really problematic here i presume.. the rate-limiter will be inresponsive if the burst is zero according to its comment docs:
```go
// The bucket is initially filled with 'burst' tokens, and refills at a rate of 'qps'.
// The maximum number of tokens in the bucket is capped at 'burst'.
```
the field should always be defaulted to `10` if it's zero but we are not applying/overriding the default value in that constructor func so the rate-limiter will be broken. i think the correct fix is to add a check of `configShallowCopy.Burst > 0` as is linked in the following, the point of shallow-copying config there is making all the typed-client share a same rate-limiter #24166.

https://github.com/kubernetes/kubernetes/blob/8f927d881507b12aacd9a8b4b6a7866c2405c7f7/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go#L139-L141

**Comment by yuyulei** [comment]
On the one hand, if `configShallowCopy.Burst==0`, the  rate-limiter will be broken. On the other hand, [apiextensions-apiserver.NewForConfig](https://github.com/kubernetes/kubernetes/blob/eaf89cfbb483f99411141f70c294e8a014e49cec/staging/src/k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/clientset.go#L54) call [client-go.RESTClientFor](https://github.com/kubernetes/client-go/blob/master/rest/config.go#L279), and The latter will check qps, burst ?= 0 and build rate-limiter. So Its no need to build rate-limiter in the begining "apiextensions-apiserver.NewForConfig".

**Comment by yuyulei** [comment]
```
if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS > 0 {
		configShallowCopy.RateLimiter = flowcontrol.NewTokenBucketRateLimiter(configShallowCopy.QPS, configShallowCopy.Burst)
	}
```
appears in many places, like: 
`k8s.io/kubernetes/staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go`
`k8s.io/kubernetes/staging/src/k8s.io/client-go/kubernetes/clientset.go`
`k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/clientset.go`
...
even `k8s.io/kubernetes/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go`

if fix https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go#L139, other places will be ok ?
or fix  it in every place? Need some suggestion.

**Comment by roycaihw** [comment]
@yuyulei Yes, to modify the generated files, you need to first fix the code generator (client-gen in this case), then re-run the code generator using either `hack/update-codegen.sh` or `make update`

I'm reading the code to see if the ratelimiter can bypass the defaulting

/ok-to-test

**Comment by roycaihw** [comment]
the code truly is problematic. The ratelimter [hangs with 0 burst](https://github.com/roycaihw/kubernetes/commit/cb6a58c702bc78d792d22dfa9d62cbb0fa209a79#diff-0dc6e090e6e256d6827a017d02bbf6e0R57)
```
E0619 13:58:29.525415  205664 main.go:57] qps: 0
E0619 13:58:29.525679  205664 main.go:58] burst: 0
E0619 13:58:29.525688  205664 main.go:59] ratelimiter: <nil>
There are 1 CRDs in the cluster
There are 1 CRDs in the cluster
...
```

```
E0619 13:58:58.032807  206343 main.go:58] qps: 5
E0619 13:58:58.033085  206343 main.go:59] burst: 0
E0619 13:58:58.033093  206343 main.go:60] ratelimiter: <nil>
hang
```

the fix https://github.com/kubernetes/kubernetes/pull/79077#issuecomment-502744804 @yue9944882 suggested looks good. 

I'm not sure if I understood #24166 though. IMO if the RateLimiter was nil, new RateLimiters will be created for each configShallowCopy in each client. If the RateLimiter wasn't nil, it would be shared by all clients without the change in #24166

**Comment by yuyulei** [comment]
@roycaihw  after reading [#24166], I think I should not remove code. And when building the shared rate-limiter, if  qps > 0 but burst =0, we may set default value to burst. If you think ok, I will fix it.

**Comment by roycaihw** [comment]
I'm okay with defaulting at early stage in the constructor to promote the shared rate limiter. @yue9944882 WDYT?

one concern is that 0-value burst works when [qps is Inf](https://github.com/kubernetes/kubernetes/blob/c898ab2043655758f07c412a98c7388e5b4dea69/vendor/golang.org/x/time/rate/rate.go#L22), but I doubt if any user accidentally depends on that behavior to "disable" rate limiting

**Comment by yue9944882** [comment]
> I'm okay with defaulting at early stage in the constructor to promote the shared rate limiter. 

mmm i feel complex about promoting the defaulting.. to make it cleaner, just checking `burst > 0` is fine i guess..

> one concern is that 0-value burst works when qps is Inf, but I doubt if any user accidentally depends on that behavior to "disable" rate limiting

it's a dummy rate-limtter when we build a client w/ `Inf` qps and `0` burst. so it doesn't matter whether they're sharing, amirght?

**Review by yue9944882 — APPROVED** [review]

**Comment by liggitt** [comment]
would it be better to return an error if QPS > 0 and burst <= 0?

**Comment by yuyulei** [comment]
@liggitt 

1. have updated code: add check when qps<0 and burst <0. 
2. two places not updated: [custom_metrics](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go#L56) and [external_metrics](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go#L38), because they were not code-generated by client-gen. So need to update them manually or ignore them?

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go [hunk 1]
```diff
@@ -51,9 +53,17 @@ func (c *Clientset) Discovery() discovery.DiscoveryInterface {
 }
 
 // NewForConfig creates a new Clientset for the given config.
+// If config's RateLimiter is not set and QPS and Burst are acceptable,
+// NewForConfig will generate a shared rate-limiter for config.
 func NewForConfig(c *rest.Config) (*Clientset, error) {
 	configShallowCopy := *c
-	if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS > 0 {
+	if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS < 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go]: we cannot tighten this validation. -1 means "do not rate limit". see test for this at https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/config/validation/validation_test.go#L50-L53

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go [hunk 2]
```diff
@@ -51,9 +53,17 @@ func (c *Clientset) Discovery() discovery.DiscoveryInterface {
...
+		return nil, fmt.Errorf("QPS is required to be greater than or equal to 0 when RateLimiter is not set")
+	}
+	if configShallowCopy.RateLimiter == nil && configShallowCopy.Burst < 0 {
+		return nil, fmt.Errorf("Burst is required to be greater than or equal to 0 when RateLimiter is not set")
+	}
+	if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS > 0 && configShallowCopy.Burst > 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go]: the case I actually wondered about erroring on QPS > 0 and Burst == 0. If that hangs today, then erroring would make sense to me (we're not causing anything that is currently working to fail, and it makes the failure more obvious)

**Comment by liggitt** [comment]
> 2\. two places not updated: [custom_metrics](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go#L56) and [external_metrics](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/metrics/pkg/client/external_metrics/client.go#L38), because they were not code-generated by client-gen. So need to update them manually or ignore them?

I'd expect to make those match

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go (hunk 2 above)
Comment (yuyulei) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go]: got it

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go (hunk 1 above)
Comment (yuyulei) [reply staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/clientset/versioned/clientset.go]: I will fix it.

**Comment by yuyulei** [comment]
@liggitt PTAL

**Comment by yuyulei** [comment]
@liggitt

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go [hunk 3]
```diff
@@ -134,9 +134,15 @@ func (c *Clientset) Discovery() $.DiscoveryInterface|raw$ {
 
 var newClientsetForConfigTemplate = `
 // NewForConfig creates a new Clientset for the given config.
+// If config's RateLimiter is not set and QPS and Burst are acceptable, 
+// NewForConfig will generate a rate limit in config, and the rate 
+// limit will be shared by other clients from same config.
 func NewForConfig(c *$.Config|raw$) (*Clientset, error) {
 	configShallowCopy := *c
 	if configShallowCopy.RateLimiter == nil && configShallowCopy.QPS > 0 {
+		if configShallowCopy.Burst == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go]: should this be checking `...Burst <= 0`?

**Comment by yuyulei** [comment]
@liggitt

File: staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go [hunk 4]
```diff
@@ -134,9 +134,15 @@ func (c *Clientset) Discovery() $.DiscoveryInterface|raw$ {
 
 var newClientsetForConfigTemplate = `
 // NewForConfig creates a new Clientset for the given config.
+// If config's RateLimiter is not set and QPS and Burst are acceptable, 
+// NewForConfig will generate a rate limit in config, and the rate 
+// limit will be shared by other clients from same config.
```
Comment (liggitt) [inline staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go]: sorry... just noticed this comment... the rate limiter is set in configShallowCopy, not in config, so other clients created from the same config will not share the ratelimiter, right?

Comment (yuyulei) [reply staging/src/k8s.io/code-generator/cmd/client-gen/generators/generator_for_clientset.go]: yes, I made a mistake,

**Comment by yuyulei** [comment]
Thanks:)

**Pull request merged**